perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,717 @@
+    auto name = trt_engine_->getBindingName(i);
+    buffer_map_[name] = i;
+    print_dim(dim, name);
+  }
+
+  return EXIT_SUCCESS;
+}
+
+int BEVDet::deserializeTRTEngine(
+  const std::string & engine_file, nvinfer1::ICudaEngine ** engine_ptr)
+{
+  std::stringstream engine_stream;
+  engine_stream.seekg(0, engine_stream.beg);
+
+  std::ifstream file(engine_file);
+  engine_stream << file.rdbuf();
+  file.close();
+
+  nvinfer1::IRuntime * runtime = nvinfer1::createInferRuntime(g_logger_);
+  if (runtime == nullptr) {
+    std::string msg("Failed to build runtime parser!");
+    g_logger_.log(nvinfer1::ILogger::Severity::kERROR, msg.c_str());
+    return EXIT_FAILURE;
+  }
+  engine_stream.seekg(0, std::ios::end);
+  const int engine_size = engine_stream.tellg();
+
+  engine_stream.seekg(0, std::ios::beg);
+  void * engine_str = malloc(engine_size);
+  engine_stream.read(reinterpret_cast<char *>(engine_str), engine_size);
+
+  nvinfer1::ICudaEngine * engine = runtime->deserializeCudaEngine(engine_str, engine_size, NULL);
+  if (engine == nullptr) {
+    std::string msg("Failed to build engine parser!");
+    g_logger_.log(nvinfer1::ILogger::Severity::kERROR, msg.c_str());
+    return EXIT_FAILURE;
+  }
+  *engine_ptr = engine;
+  for (int bi = 0; bi < engine->getNbBindings(); bi++) {
+    if (engine->bindingIsInput(bi) == true) {
+      RCLCPP_INFO(
+        rclcpp::get_logger("BEVDet"), "Binding %d (%s): Input.", bi, engine->getBindingName(bi));
+    } else {
+      RCLCPP_INFO(
+        rclcpp::get_logger("BEVDet"), "Binding %d (%s): Output.", bi, engine->getBindingName(bi));
+    }
+  }
+  return EXIT_SUCCESS;
+}
+
+void BEVDet::getAdjBEVFeature(
+  const std::string & curr_scene_token, const Eigen::Quaternion<float> & ego2global_rot,
+  const Eigen::Translation3f & ego2global_trans)
+{
+  int flag = 1;
+  if (adj_frame_ptr_->lastScenesToken() != curr_scene_token) {
+    adj_frame_ptr_->reset();
+    flag = 0;
+  }
+
+  // the smaller the idx, the newer th adj_bevfeat
+  for (int i = 0; i < adj_num_; i++) {
+    const void * adj_buffer = adj_frame_ptr_->getFrameBuffer(i);
+
+    size_t buf_size = trt_buffer_sizes_[buffer_map_["adj_feats"]] / adj_num_;
+
+    CHECK_CUDA(cudaMemcpy(
+      reinterpret_cast<char *>(trt_buffer_dev_[buffer_map_["adj_feats"]]) + i * buf_size,
+      adj_buffer, buf_size, cudaMemcpyDeviceToDevice));
+
+    Eigen::Quaternion<float> adj_ego2global_rot;
+    Eigen::Translation3f adj_ego2global_trans;
+    adj_frame_ptr_->getEgo2Global(i, adj_ego2global_rot, adj_ego2global_trans);
+
+    getCurr2AdjTransform(
+      ego2global_rot, adj_ego2global_rot, ego2global_trans, adj_ego2global_trans,
+      reinterpret_cast<float *>(trt_buffer_dev_[buffer_map_["transforms"]]) + i * transform_size_);
+  }
+  CHECK_CUDA(cudaMemcpy(
+    trt_buffer_dev_[buffer_map_["flag"]], &flag, trt_buffer_sizes_[buffer_map_["flag"]],
+    cudaMemcpyHostToDevice));
+}
+
+void BEVDet::getCurr2AdjTransform(
```

Comment:
Namespaces: Wrap the function `getCurr2AdjTransform` within the  `autoware_tensorrt_bevdet` namespace.